test/performance/odp_dmafwd.c

```diff
@@ -0,0 +1,973 @@
+	pool_param.buf.size = sizeof(transfer_t);
+	pool_param.type = ODP_POOL_BUFFER;
+	config->trs_pool = odp_pool_create(PROG_NAME "_dma_trs", &pool_param);
+
+	if (config->trs_pool == ODP_POOL_INVALID) {
+		ODPH_ERR("Error creating DMA transfer tracking pool\n");
+		return false;
+	}
+
+	pool_param.pkt.seg_len = config->pkt_len;
+	pool_param.pkt.len = config->pkt_len;
+	pool_param.pkt.num = config->num_pkts;
+	pool_param.type = ODP_POOL_PACKET;
+	config->copy_pool = odp_pool_create(PROG_NAME "_dma_copy", &pool_param);
+
+	if (config->copy_pool == ODP_POOL_INVALID) {
+		ODPH_ERR("Error creating DMA packet copy pool\n");
+		return false;
+	}
+
+	for (int i = 0; i < config->num_thrs; ++i) {
+		thr = &config->thread_config[i];
+		thr->dma_handle = odp_dma_create(PROG_NAME "_dma", &dma_params);
+
+		if (thr->dma_handle == ODP_DMA_INVALID) {
+			ODPH_ERR("Error creating DMA session\n");
+			return false;
+		}
+
+		odp_dma_pool_param_init(&compl_pool_param);
+		compl_pool_param.num = config->num_pkts;
+		thr->compl_pool = odp_dma_pool_create(PROG_NAME "_dma_compl", &compl_pool_param);
+
+		if (thr->compl_pool == ODP_POOL_INVALID) {
+			ODPH_ERR("Error creating DMA event completion pool\n");
+			return false;
+		}
+
+		thr->copy_pool = config->copy_pool;
+		thr->trs_pool = config->trs_pool;
+		odp_queue_param_init(&queue_param);
+		queue_param.type = ODP_QUEUE_TYPE_SCHED;
+		queue_param.sched.sync = ODP_SCHED_SYNC_PARALLEL;
+		queue_param.sched.prio = odp_schedule_max_prio();
+		thr->compl_q = odp_queue_create(PROG_NAME "_dma_compl", &queue_param);
+
+		if (thr->compl_q == ODP_QUEUE_INVALID) {
+			ODPH_ERR("Error creating DMA completion queue\n");
+			return false;
+		}
+	}
+
+	config->copy_fn = dma_copy_and_send_packets;
+
+	return true;
+}
+
+static odp_bool_t setup_pktios(prog_config_t *config)
+{
+	odp_pool_param_t pool_param;
+	pktio_t *pktio;
+	odp_pktio_param_t pktio_param;
+	odp_pktin_queue_param_t pktin_param;
+	odp_pktout_queue_param_t pktout_param;
+
+	odp_pool_param_init(&pool_param);
+	pool_param.pkt.seg_len = config->pkt_len;
+	pool_param.pkt.len = config->pkt_len;
+	pool_param.pkt.num = config->num_pkts;
+	pool_param.type = ODP_POOL_PACKET;
+	config->pktio_pool = odp_pool_create(PROG_NAME, &pool_param);
+
+	if (config->pktio_pool == ODP_POOL_INVALID) {
+		ODPH_ERR("Error creating packet I/O pool\n");
+		return false;
+	}
+
+	for (uint32_t i = 0U; i < config->num_ifs; ++i) {
+		pktio = &config->pktios[i];
+		odp_pktio_param_init(&pktio_param);
+		pktio_param.in_mode = ODP_PKTIN_MODE_SCHED;
+		pktio_param.out_mode = ODP_PKTOUT_MODE_DIRECT;
+		pktio->handle = odp_pktio_open(pktio->name, config->pktio_pool, &pktio_param);
+
+		if (pktio->handle == ODP_PKTIO_INVALID) {
+			ODPH_ERR("Error opening packet I/O (%s)\n", pktio->name);
+			return false;
+		}
+
+		odp_pktin_queue_param_init(&pktin_param);
+		pktin_param.queue_param.sched.prio = odp_schedule_default_prio();
+
+		if (odp_pktin_queue_config(pktio->handle, &pktin_param) < 0) {
```

Comment:
Multiple input/outout queues could be used. Can be a future improvement.

Reply:
V2.